[state_dict] Add cpu_only and ranks_only support for _gather_state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112837
* #112885
* __->__ #112836

Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

fegin committed:
[state_dict] Add cpu_only and ranks_only support for _gather_state_dict

Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

wz337 commented on torch/distributed/checkpoint/_state_dict_utils.py:
```diff
@@ -49,13 +49,34 @@ def _all_gather_sharded_tensor(
+    Args:
+        state_dict (Dict[str, Any]): the target sharded state_dict.
+        pg (Optional[dist.ProcessGroup]): the process group that is used to
+            gather ShardedTensor.
```
nit: gather DTensor or ShardedTensor.

wz337 commented on torch/distributed/checkpoint/_state_dict_utils.py:
```diff
@@ -49,13 +49,34 @@ def _all_gather_sharded_tensor(
+        pg (Optional[dist.ProcessGroup]): the process group that is used to
+            gather ShardedTensor.
+        device: (Optional[torch.device]): the device that is used to
+            perform allgather for ShardedTensor.
```
nit: gather DTensor or ShardedTensor.

fegin replied:
ShardedTensor only. DTensor uses DeviceMesh to gather the tensors.

wz337 commented on torch/distributed/checkpoint/_state_dict_utils.py:
```diff
@@ -86,7 +107,11 @@ def _gather_state_dict(
+            value = value.to(cpu_device)
 
-        new_state_dict[key] = value
+        if not cpu_offload or len(ranks_only) == 0 or dist.get_rank(pg) in ranks_only:
```
Is `dist.get_rank(pg) in ranks_only` along not sufficient?

fegin replied:
ranks_only can be empty (the default value). But it seems that we should not restrict this function to be cpu_offload only.

wz337 replied:
I see. Makes sense!

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]

LucasLLC reviewed (approved):
lgtm!

fegin committed:
Update on "[state_dict] Add cpu_only and ranks_only support for _gather_state_dict"


Add cpu_only and ranks_only support for _gather_state_dict

Differential Revision: [D50962980](https://our.internmc.facebook.com/intern/diff/D50962980/)

[ghstack-poisoned]
